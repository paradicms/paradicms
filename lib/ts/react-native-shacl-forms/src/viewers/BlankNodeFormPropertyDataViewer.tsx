import {FormNodeData, FormPropertyData} from "@paradicms/shacl-forms";
import * as React from "react";
import {useState} from "react";
import {View} from "react-native";
import {ListItem, Text, useTheme} from "@rneui/themed";
import {FormNodeDataViewer} from "./FormNodeDataViewer";
import {FormPropertyDataViewerFactory} from "./FormPropertyDataViewerFactory";
import {BlankNode, NamedNode} from "@rdfjs/types";
import {IconFactory} from "../IconFactory";
import {getRdfNodeLabel} from "@paradicms/rdf";

export const BlankNodeFormPropertyDataViewer: React.FunctionComponent<{
  formPropertyData: FormPropertyData;
  formPropertyDataViewerFactory: FormPropertyDataViewerFactory;
  iconFactory: IconFactory;
}> = ({formPropertyData, formPropertyDataViewerFactory, iconFactory}) => {
  const [expandedValueI, setExpandedValueI] = useState<number>(-1);
  const {theme} = useTheme();

  const values = formPropertyData.values.filter(
    value => value.termType === "BlankNode"
  );
  if (values.length === 0) {
    console.info(`form property ${formPropertyData.id} has no values`);
    return null;
  }

  const nodeShapes = formPropertyData.shape.nodeShapes;
  if (nodeShapes.length === 0) {
    console.info(`form property ${formPropertyData.id} has no node shapes`);
    return null;
  } else if (nodeShapes.length > 1) {
    console.info(
      `form property ${formPropertyData.id} has ${nodeShapes.length} node shapes`
    );
    return null;
  }
  const nodeShape = nodeShapes[0];

  return (
    <View>
      {values.map((value, valueI) => (
        <ListItem.Accordion
          content={
            <Text>
              {getRdfNodeLabel({
                dataset: formPropertyData.dataGraph,
                node: value as BlankNode | NamedNode,
              }) ?? `${formPropertyData.label} ${valueI + 1}`}
            </Text>
          }
          expandIcon={iconFactory({name: "chevron-up"})}
          isExpanded={expandedValueI === valueI}
          icon={iconFactory({name: "chevron-down"})}
          key={valueI + 1}
          onPress={() =>
            setExpandedValueI(expandedValueI !== valueI ? valueI : -1)
          }
        >
          <View style={{marginLeft: theme.spacing.xl}}>
            <FormNodeDataViewer
              formNodeData={
                new FormNodeData({
                  dataGraph: formPropertyData.dataGraph,
                  dataGraphNode: value as BlankNode | NamedNode,
                  shape: nodeShape,
                })
              }
              formPropertyDataViewerFactory={formPropertyDataViewerFactory}
              iconFactory={iconFactory}
            />
          </View>
        </ListItem.Accordion>
      ))}
    </View>
  );
};
