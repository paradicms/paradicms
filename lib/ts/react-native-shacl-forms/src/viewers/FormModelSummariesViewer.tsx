import * as React from "react";
import {FormModelSummary} from "@paradicms/shacl-forms";
import {View} from "react-native";
import {Button, ListItem} from "@rneui/themed";

interface FormModelSummariesViewerProps<
  FormModelSummaryT extends FormModelSummary
> {
  formModelSummaries: readonly FormModelSummaryT[];
  onSelect: (
    formModelSummary: FormModelSummaryT,
    formModelSummaryIndex: number
  ) => void;
}

export const FormModelSummariesViewer = <
  FormModelSummaryT extends FormModelSummary
>(
  props: FormModelSummariesViewerProps<FormModelSummaryT>
) => {
  const {formModelSummaries, onSelect} = props;
  return (
    <View>
      {formModelSummaries.map((formModelSummary, formModelSummaryIndex) => (
        <ListItem key={formModelSummary.id}>
          <Button
            onPress={() => onSelect(formModelSummary, formModelSummaryIndex)}
            title={formModelSummary.label}
          />
        </ListItem>
      ))}
    </View>
  );
};
