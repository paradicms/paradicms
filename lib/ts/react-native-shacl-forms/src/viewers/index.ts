export * from "./createFormPropertyDataViewer";
export * from "./FormDataViewer";
export * from "./FormModelSummariesViewer";
export * from "./FormNodeDataViewer";
export * from "./FormNodeTypeDataViewer";
