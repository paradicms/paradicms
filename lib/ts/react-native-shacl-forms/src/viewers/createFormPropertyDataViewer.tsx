import * as React from "react";
import {dash} from "@paradicms/vocabularies";
import {TextFormPropertyDataViewer} from "./TextFormPropertyDataViewer";
import {BlankNodeFormPropertyDataViewer} from "./BlankNodeFormPropertyDataViewer";
import {FormPropertyDataViewerFactory} from "./FormPropertyDataViewerFactory";

export const createFormPropertyDataViewer: FormPropertyDataViewerFactory = ({
  formPropertyData,
  iconFactory,
}) => {
  const viewer = formPropertyData.viewer;
  if (!viewer) {
    console.info(`form property ${formPropertyData.id} has no viewer`);
    return null;
  }

  if (viewer.equals(dash.BlankNodeViewer)) {
    return (
      <BlankNodeFormPropertyDataViewer
        formPropertyData={formPropertyData}
        formPropertyDataViewerFactory={createFormPropertyDataViewer}
        iconFactory={iconFactory}
      />
    );
  } else if (viewer.equals(dash.LiteralViewer)) {
    return <TextFormPropertyDataViewer formPropertyData={formPropertyData} />;
  } else {
    console.info(
      `form property ${formPropertyData.id} has unsupported viewer ${viewer.value}`
    );
    return null;
  }
};
