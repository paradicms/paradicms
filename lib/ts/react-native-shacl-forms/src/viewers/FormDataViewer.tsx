import * as React from "react";
import {FormData, FormNodeTypeData} from "@paradicms/shacl-forms";
import {FormModelSummariesViewer} from "./FormModelSummariesViewer";

export const FormDataViewer: React.FunctionComponent<{
  formData: FormData;
  onSelectFormNodeTypeData: (formNodeTypeData: FormNodeTypeData) => void;
}> = ({formData, onSelectFormNodeTypeData}) => {
  return (
    <FormModelSummariesViewer
      formModelSummaries={formData.nodeTypes}
      onSelect={onSelectFormNodeTypeData}
    />
  );
};
