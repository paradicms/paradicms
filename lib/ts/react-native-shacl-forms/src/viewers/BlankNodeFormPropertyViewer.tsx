import {FormNode, FormProperty} from "@paradicms/shacl-forms";
import * as React from "react";
import {StyleSheet, View} from "react-native";
import {ListItem, Text} from "@rneui/themed";
import {FormNodeViewer} from "./FormNodeViewer";
import {BlankNode, NamedNode} from "@rdfjs/types";
import {FormPropertyViewerFactory} from "./FormPropertyViewerFactory";

export const BlankNodeFormPropertyViewer: React.FunctionComponent<{
  formProperty: FormProperty;
  formPropertyViewerFactory: FormPropertyViewerFactory;
}> = ({formProperty, formPropertyViewerFactory}) => {
  const values = formProperty.values.filter(
    value => value.termType === "BlankNode"
  );
  if (values.length === 0) {
    console.info(`form property ${formProperty.id} has no values`);
    return null;
  }

  const nodeShapes = formProperty.shape.nodeShapes;
  if (nodeShapes.length === 0) {
    console.info(`form property ${formProperty.id} has no node shapes`);
    return null;
  } else if (nodeShapes.length > 1) {
    console.info(
      `form property ${formProperty.id} has ${nodeShapes.length} node shapes`
    );
    return null;
  }
  const nodeShape = nodeShapes[0];

  return (
    <View>
      <View style={styles.labelContainer}>
        <Text style={styles.label}>{formProperty.label}</Text>
      </View>
      <View>
        {values.map((value, valueI) => {
          const formNode = new FormNode({
            dataGraph: formProperty.dataGraph,
            dataGraphNode: value as BlankNode | NamedNode,
            shape: nodeShape,
            shapesGraph: formProperty.shapesGraph,
          });
          return (
            <ListItem key={valueI}>
              <FormNodeViewer
                formNode={formNode}
                formPropertyViewerFactory={formPropertyViewerFactory}
              />
            </ListItem>
          );
        })}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  label: {
    fontWeight: "bold",
  },
  labelContainer: {
    alignItems: "center",
  },
});
