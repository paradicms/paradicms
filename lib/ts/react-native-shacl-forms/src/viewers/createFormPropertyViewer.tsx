import * as React from "react";
import {dash} from "@paradicms/vocabularies";
import {FormProperty} from "@paradicms/shacl-forms";
import {TextFormPropertyViewer} from "./TextFormPropertyViewer";
import {BlankNodeFormPropertyViewer} from "./BlankNodeFormPropertyViewer";
import {FormPropertyViewerFactory} from "./FormPropertyViewerFactory";
import {IconFactory} from "../IconFactory";

export const createFormPropertyViewer: FormPropertyViewerFactory = (kwds: {
  formProperty: FormProperty;
  iconFactory: IconFactory;
}): React.ReactElement | null => {
  const {formProperty, iconFactory} = kwds;

  const viewer = formProperty.viewer;
  if (!viewer) {
    console.info(`form property ${formProperty.id} has no viewer`);
    return null;
  }

  if (viewer.equals(dash.BlankNodeViewer)) {
    return (
      <BlankNodeFormPropertyViewer
        formProperty={formProperty}
        formPropertyViewerFactory={createFormPropertyViewer}
        iconFactory={iconFactory}
      />
    );
  } else if (viewer.equals(dash.LiteralViewer)) {
    return <TextFormPropertyViewer formProperty={formProperty} />;
  } else {
    console.info(
      `form property ${formProperty.id} has unsupported viewer ${viewer.value}`
    );
    return null;
  }
};
