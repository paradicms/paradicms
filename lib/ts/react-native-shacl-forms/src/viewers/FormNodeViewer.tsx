import * as React from "react";
import {FormNode} from "@paradicms/shacl-forms";
import {StyleSheet, View} from "react-native";
import {createFormPropertyViewer} from "./createFormPropertyViewer";

export const FormNodeViewer: React.FunctionComponent<{
  formNode: FormNode;
}> = ({formNode}) => (
  <View>
    {formNode.properties.map(formProperty => (
      <View key={formProperty.id} style={styles.formPropertyViewerContainer}>
        {createFormPropertyViewer({formProperty})}
      </View>
    ))}
  </View>
);

const styles = StyleSheet.create({
  formPropertyViewerContainer: {
    marginBottom: 25,
  },
});
