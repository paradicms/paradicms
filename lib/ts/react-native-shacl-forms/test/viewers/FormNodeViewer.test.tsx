import * as React from "react";
import * as renderer from "react-test-renderer";
import {testForm} from "../testForm";
import {FormNodeViewer} from "../../src/viewers/FormNodeViewer";

test("renders correctly", () => {
  const formNode = testForm().nodeTypes[0].nodes[0];
  const tree = renderer.create(<FormNodeViewer formNode={formNode} />).toJSON();
  expect(tree).toMatchSnapshot();
});
