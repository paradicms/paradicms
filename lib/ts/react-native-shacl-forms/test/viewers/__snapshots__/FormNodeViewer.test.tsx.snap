// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      {
        "marginBottom": 25,
      }
    }
  />
  <View
    style={
      {
        "marginBottom": 25,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontWeight": "bold",
            }
          }
        >
          http://schema.org/gender
        </Text>
      </View>
      <View
        style={
          {
            "marginTop": 5,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
            }
          }
        >
          male
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 25,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontWeight": "bold",
            }
          }
        >
          given name
        </Text>
      </View>
      <View
        style={
          {
            "marginTop": 5,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
            }
          }
        >
          Minor
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 25,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontWeight": "bold",
            }
          }
        >
          http://schema.org/birthDate
        </Text>
      </View>
      <View
        style={
          {
            "marginTop": 5,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
            }
          }
        >
          2022-11-19
        </Text>
      </View>
    </View>
  </View>
</View>
`;
